service/migration_manager.hh

```diff
@@ -80,6 +80,8 @@ public:
 
     migration_notifier& get_notifier() { return _notifier; }
     const migration_notifier& get_notifier() const { return _notifier; }
+    service::storage_proxy& get_storage_proxy() { return _storage_proxy; }
+    const service::storage_proxy& get_storage_proxy() const { return _storage_proxy; }
```

Comment:
@xemul as the person who is untangling our dependency mess, is this getter fine?

Reply:
@xemul, in case I didn't explain the raison d'etre of this new getter well enough in the commit message, I'll try again here: What happens is that I have code that gets a `migration_manager` and needs a `data_dictionary`. The migration_manager doesn't have a data_dictionary in it, but when its implementation needs one it uses the `storage_proxy` it holds and there is a data_dictionary in it. As a **user** of migration_manager, I now want to do exactly the same, and this is why I needed this getter.
If you think there's a better way for a holder of migration_manager to find a data_dictionary, I can use that way as well.

Reply:
>  No, that's the entire point of this PR - we need the modify_tags operation to be atomic when facing concurrent schema operations, so it can't take the "schema" object which was fixed when got it (and might be outdated if a parallel schema change happened!), we needs to get the current schema for the given ks/cf after taking the schema-modification lock. So we need a data_dictionary.

I see

> As I noted above, mm also needs, internally a data_dictionary, and it uses exactly the same trick I used (get the data_dictionary from the storage proxy it holds). But here I need it in code using mm, not inside mm...

Well. I've no objections against exporting proxy from mm other than described above. I guess it's an overcompensation since the times I met some long chain to get ... db::config? from. It was like `auto& cfg = storage_sevice.get_query_processor().local().proxy().local().get_db().local().get_config()`. Apparently this chain accumulated over time just because there was a way to get closer to the config by using exporter from different services.

Last chance not to export proxy from mm -- how about passing `noncopyable_function<schema_ptr()> get_schema` as an argument instead of `sstring ks, sstring cf` pair? The caller is executor which can initialize this argument with lambda and `sharded<storage_proxy>&` capture to get data-dictionary from.

Reply:
The function trick is not only "ugly" (what a bizarre API this would be...) I'm worried it will be difficult to do correctly because another important complication in this code is that the code needs to run the schema modification on a different core (always core 0), so the code (which you can see in this patch) automatically gets the right storage_proxy for core 0, but it doesn't get other things for core 0, even if executor had them. Maybe I could capture the sharded storage proxy (as you suggested) and use `local()` to use it on core 0, but this would be very ugly...

Another alternative thing I could do is not to have in migration_manager a getter for storage_proxy, but instead, a getter for data_dictionary (which is the only thing I needed anyway). The rationale can be that migration_manager already has various methods which take ks_name and cf_name, so it **must** have access to data_dictionary already, so if it must have such access, it might as well give its users access to it as well.

Reply:
I don't mind this change - it's usually easy to unravel it if it turns out it's wrong.

I agree it's better to pass the dependency directly.